### ISSUE TITLE
Can't game on Windows 98 SE unless sbtype = none

### ISSUE THREAD

**Issue by brunocastello** [joncampbell123/dosbox-x#2689]
**Describe the bug**
I cannot run any game at all on Windows 98 SE, unless I set sbtype to "none". FIFA 98 RTWC, FIFA 99, Need For Speed II SE, all them cannot run with sound on, even when I try to play them in software mode instead of glide pass-through.

NFS2SE with both software and 3dfx crashes when the race is about to start, during the intro flying into the player's car.
FIFA 98 RTWC and FIFA 99 both crash when the match kick-off is about to start, no matter if 3dfx is enabled or not.

**To Reproduce**
Play any of the three games above (a demo is enough) under Windows 98 SE, using sbtype = sb16 or sb16vibra.

**Expected behavior**
Games to run flawlessly with sound enabled.

**Environment (please complete the following information):**
 - Operating system: macOS 11.4 Big Sur [Intel]
 - DOSBox-X release version: SDL1.2 0.83.15 (The problem is persistent since several versions before)
 - Used configuration: attached below

**Additional context**
I am running a build without avcodec and fluidsynth. I do not want all the 93893 dependencies both require to run. In a Mac where hard disk space is a premium, this is needed. I just need networking and the gaming features; I do not want extra sound features (fluidsynth) or to do any screen recording (avcodec) stuff. However, even using the standard build with them, the problem persists.

The computer has [this configuration](https://everymac.com/systems/apple/macbook_pro/specs/macbook-pro-core-i7-2.8-13-late-2013-retina-display-specs.html) and has been upgraded to 16GB RAM at the time of purchase.

Attached is my dosbox configuration file, pretty similar to the one from the wiki guide for Windows 98 installation.
[dosbox.txt](https://github.com/joncampbell123/dosbox-x/files/6837267/dosbox.txt)

**Comment by Wengier** [comment]
I have not yet tried installing Windows 98 SE on macOS. I wonder if this issue is specific to macOS?

**Comment by brunocastello** [comment]
If it is specific, I'd love to hear.  I have noticed that DOSBox-X gives wrong IRQ, DMA and HDMA to the SB16 when I reach the Windows 98 Desktop. But fixing that issue did not fix my problem. Here's the thing:

I've set I7 D1 H5 for sb16 in dosbox.conf. yet when checking the assigned values from DOSBox sound menu, it's 2, 3 and 5 (I can't remember now, I'm at work, but basically it's not what I have assigned). Then I removed it from dosbox.conf and manually assigned them through Device Manager in Windows 98SE, as well as set the blaster environment in autoexec.bat. Now DOSBox-X reports the right configuration, however, the games still crash.

Since it crashes with both glide and software mode when sound is enabled, I already discarded any issue with OpenGlide; I have tried with both kjliew's and almeath's OpenGlide implementations as well.

Could it be a SDL audio issue? Does SDL even deal with it?

**Comment by Wengier** [comment]
@brunocastello It is interesting to hear that the sound menu shows a different IRQ than the one assigned in the config file. Does this issue happen when you are in DOSBox-X' integrated DOS? Or only when you boot to Windows 98 desktop?

As for SDL, it *might* be related. Have you tried the SDL2 build and see if there are any changes?

**Comment by brunocastello** [comment]
@Wengier happens when I boot to Windows 98. I haven’t checked when using integrated DOS, because I have never used. But shows correct one when I force them through Windows 98 Device Manager.

SDL: I didn’t even try SDL2 as far as booting into Windows 98, because the SDL2 build has 2 problems for me, the yellow screen (no matter what video driver I use, it is always yellow) and OpenGlide works only with SDL12. So without OpenGlide I couldn’t test these games in SDL2 build...

**Comment by Wengier** [comment]
@brunocastello The main purpose for trying the SDL build is not to actually use it as the main build, but to try to isolate the issue. Whether it works there may help us find out the cause of the problem. Similar for the IRQ assignment.

**Comment by brunocastello** [comment]
OK, I will try to find some time next night to try it out. I was out for dinner this night.

**Comment by brunocastello** [comment]
@Wengier compiling a SDL2 build to test as we speak... btw the compiler returned this message:
`configure: WARNING: unrecognized options: --disable-video-x11-xv`
Pretty sure something was overlooked there?

**Comment by stephenbonar** [comment]
Ah, I'm having the same issue! I couldn't understand why so many games were crashing in Win98SE until I saw this issue. In my case, the games I was trying were not 3D accelerated (Heroes of Might and Magic II, Might and Magic VI), but as soon as I set sbtype=none, the games are stable. It did seem like certain sounds were possibly triggering the crashes. I just tried changing the config on the Apple M1 version of Dosbox-x (0.83.15), but I had the same problem on other platforms (Windows x86_64). I have yet to see if this fixes the crashes on the x86_64 builds.

**Comment by stephenbonar** [comment]
In the issue I linked to in my previous comment (#2621), the author discovered that changing the CPU type from pentium_mmx to pentium resolved the issue. Sure enough, I tried it on my M1 Mac and it indeed resolved the issue for me as well (I believe the author was on x86_64, but it resolved it on both platforms). @brunocastello I would try this and see if it fixes the issue with your games too. If it does, I would suspect there's an issue with sound involving the pentium_mmx cpu type. @Wengier , any thoughts on that possibility?

**Comment by brunocastello** [comment]
@stephenbonar actually I have tried it before, and no, it does not solve my issue. No matter what cpu I set for Windows 98SE, all games will crash if I attempt to play with sound on. Without sound I can play them all.

I have tried pentium mmx, pentium pro, pentium, 486, and even set it to auto. no luck.

**Comment by brunocastello** [comment]
The problem is still persistent with the most recent builds released, both SDL1 and SDL2.

**Comment by brunocastello** [comment]
@Wengier I think I have made some little progress. I went to dxdiag in Windows 98 SE and turned off Hardware sound acceleration. This time I was able to play for a couple of seconds before it crashed to the desktop again.

NFS2SE: I nearly managed to complete a lap before CTD
FIFA 98: Managed to score a goal after kick-off before CTD
FIFA 99: CTD right on kick-off (no change of behavior)

There is definitely a sound problem and it is related to the game's speed.

**Comment by brunocastello** [comment]
Version 0.83.17, still unable to game with sound on. Games crash on Windows 98SE when there is sound.

**Comment by brunocastello** [comment]
Anyone with any suggestion about what else I can do to work around this issue? It has been there since July

**Comment by Allofich** [comment]
Your initial post says "DOSBox-X release version: SDL1.2 0.83.15 (The problem is persistent since several versions before)". Were you able to play those games without issue on some version in the past? If you have time to test, working back through the earlier versions one-by-one to find where the problem occurred would probably be a significant step to narrowing down the issue.

On the other hand if no version works no matter how back you try, it might be an issue that was always there or a problem with your settings.

Also, one thing I would also try is setting your SB16 IRQ to 5 in the .conf file (or set it to 0, in which case it DOSBox-X will automatically set it to the default value for the sound card, 5 in this case). As I understand, 5 was the default value used by SB16 cards, while 7 was for an earlier card.

**Comment by rderooy** [comment]
The original soundblaster cards indeed defaulted to IRQ7, as on XT class machines IRQ5 may be used by the harddisk controller.

IRQ7 was always a problem as it conflicts with LPT1. But few DOS programs ever used the IRQ for parallel communication, so typically you where fine, and even if they did, you would normally not have sound at the same time.

The problems started with OS/2, as you could have sound and printing simultaneous (and later with Win95 and WinNT). I remember there was at one point an update to the OS/2 printing driver to (optionally) use pulling mode, which results in higher CPU load, to prevent issues with a sound card on IRQ7.

From what I recall, the change to IRQ5 happened at the time of the SB Pro 2.

**Comment by brunocastello** [comment]
The problem predates way long before, since 0.83.11. And yes, I have tried every version released after that (and before that) back when I first reported it.

I am using the exact same settings as seen in [Windows 98 Wiki Guide](https://dosbox-x.com/wiki/Guide%3AInstalling-Windows-98#_dosbox_x_config_file). No fancy extra settings. Also tried both IRQ5 and IRQ7. Also tried all sorts of available cpus (auto, 486, pentium, pentium_mmx...).

No idea why, the games can only run when there is no sound blaster emuation. When there is, they all crash: FIFA 98 RTWC, FIFA 99, NFS2SE. I am at a complete total loss.

**Comment by brunocastello** [comment]
Tonight I tried the Windows version of DOSBox-X using a build from @Allofich (from another issue here) and in a Windows 10 Pro VMware Machine.

Turns out I have the same problem there: sbtype=sb16 and crashes, sbtype=none and I am able to play everything but without sound.

So the problem must be something with Sound Blaster inside Windows 98 that I cannot seem to debug nor find out why. I have the latest SB16 drivers as recommended from the Win98 Install Guide, as well as DirectX 9.0c.

**Comment by brunocastello** [comment]
Another night, another attempt. This time again on macOS SDL1 version with a fresh install of Windows 98 SE.

Post install: VBEMP 9x driver, DirectX 9.0c (DEC 2006) and last voodoo driver. Installed Need For Speed II SE and tried to run the 3Dfx version. Since it's a sound issue, I did not update the Sound Blaster 16 drivers this time to try and find out if the drivers are a part of the problem. Well, they're not. As soon as the race is about to start, the game crashes to the desktop.

Configuration file (renamed to .txt because GitHub wouldn't allow me to attach):
[dosbox.txt](https://github.com/joncampbell123/dosbox-x/files/7172172/dosbox.txt)

**Comment by brunocastello** [comment]
I've also noticed that I do not have the"SB16 MIDI OUT [330]" instrument listed on multimedia properties, only Creative Music Synth, "MPU 401 ..." and something like "Roland OPL3 Synth...". Don't think it has something to do, because NFS2SE, FIFA 9X aren't games with MIDI sounds AFAIK?

**Comment by rderooy** [comment]
The names of the entries in Multipedia Properties is dependent on the exact Windows release (or SB driver version) installed.

"SB16 MIDI OUT [330]" and "MPU 401..." are equivalent. I don't know where this Roland OPL3 Synth entry is from. You must have installed something as I don't recall seeing it with my testing.

Did you ever try if you get crashes with other sound card options, like a SB Pro 2?

**Comment by brunocastello** [comment]
> The names of the entries in Multipedia Properties is dependent on the exact Windows release (or SB driver version) installed.

Windows 98 SE fully updated.

> 
> "SB16 MIDI OUT [330]" and "MPU 401..." are equivalent. I don't know where this Roland OPL3 Synth entry is from. You must have installed something as I don't recall seeing it with my testing.
> 

Well, I did not. I think this comes from the oplemu setting in configuration file? This device will probably disappear if i set oplemu to none, I think.

> Did you ever try if you get crashes with other sound card options, like a SB Pro 2?

Now that you've said it... actually, no. Might try it tomorrow, it's late for me now and I haven't been sleeping well for weeks, staying late and trying to improve my QEMU/DOSBox VMs. However, I think that I need to repeat that the QEMU VM machine also has a SB16 installed, and does not have these issues. I just do not use that VM because only DOSBox-X has a proper 3Dfx emulation/Glide passthrough.

With that in mind, I'd politely ask for a revision on the SB16 card for DOSBox-x, to see if something wasn't missed or misinterpreted? Can we also check it? Because I have this issue since June/July builds.

**Comment by rderooy** [comment]
@brunocastello I do think there is a something missing in the SB16 emulation or a bug, which is causing these problems. Having said that, it requires someone with the technical knowhow, skills **and** time to fix it.

I'm not that person, and from what I have seen both @joncampbell123 and @Wengier have a lack of time right now.

**Comment by brunocastello** [comment]
> @brunocastello I do think there is a something missing in the SB16 emulation or a bug, which is causing these problems. Having said that, it requires someone with the technical knowhow, skills **and** time to fix it.
> 
> I'm not that person, and from what I have seen both @joncampbell123 and @Wengier have a lack of time right now.

I just came here to add that even the Windows version had the same problem under a VMware Windows 10 virtual machine. I used the one from @Allofich , where he fixed another issue related to img mount.

**Comment by brunocastello** [comment]
> Did you ever try if you get crashes with other sound card options, like a SB Pro 2?

Just tried it right now. As I expected, it crashed to desktop seconds before the green light for a NFS2SE race.

**Comment by brunocastello** [comment]
@Wengier and @joncampbell123 if that helps, I logged tonight an attempt to run Need For Speed II SE on Windows 98SE, you can see from this attached log that there is something happening with the SB16.

This attempt was with pentium (no mmx), cycles max, dynamic x86 core, glide passthrough enabled this time as well.

I have tried both IRQ 5 and IRQ 7, both had same result.

[dosbox.log](https://github.com/joncampbell123/dosbox-x/files/7233516/dosbox.log)

**Comment by brunocastello** [comment]
Tonight, listening more closely, I noticed that the sound is lagging a bit. Not only the Windows start sound (tadaaaa) is longer, also when I am scrolling through the NFS2SE menu options, the sound effect that happens when you select or go over an option, is also delayed. The sound certainly is not up to the emulation speed, which would explain (partially) the problem, even though I am using the best pentium cpu possible, dynamic core and cycles set to MAX.

The second part of the problem is the output when I try dxdiag to test the sound. dxdiag says it will use software buffering because my "sound card" does not have hardware buffering and it is also shown in dxdiag as an emulated SB16:

> "Your sound card does not support hardware buffering. Sounds will only play back from software buffers."

So the problem is simply that the sound in dosbox can't keep up with the emulation speed. I don't know how to work around this. Not sure if fixing the cycles to a value would help anyway, I've tried lower values (~30000, ~40000) and higher values (~60000, ~70000, ~80000) without luck.

FIFA 9X will just freeze the VM, where as NFS2SE just CTD.

Using NFS2SE as an example, when I manage to start a race before a CTD, sound appears to be matching the speeds and what is happening on screen (car tyre sounds, engine) but all of sudden there is a CTD. I do not get that problem with QEMU.

**Comment by brunocastello** [comment]
Tonight I noticed that my sound volume was set to half (50%) in Windows 98SE. I decided to set it to max (100%) moving the slider and after setting it, DOSBox-X just quit from nowhere. Weird.

**Comment by brunocastello** [comment]
I've collected a few other users issues with sb16 emulation for reference. I've also tested with cputype=pentium instead of mmx, and got the same unfortunate results. And I've also tested a plethora of different sblaster configurations, no luck either. Also tested with sbpro2. In all scenarios, FIFA 9X and NFS2SE both crashes at random times before gameplay actually starts.

#2929 is a clear copy of my situation here; #2964 appears to be the same issue, but OP did not test without sound yet; #470 is a very old 2018 issue, but reflects that the issue exists since then. I had to move my VM to QEMU, in order to be able to play and enjoy the games, although without 3dfx voodoo emulation, which would be great. However, I am still adamant that if this issue is fixed, DOSBox-X is the perfect home for these particular games (and others), because it is easier to use. Add to that the new experimental CPU, and a future M1 (or M1X) Mac that I am planning to buy, and things could be so much better.

**Comment by brunocastello** [comment]
In my case I tried all scenarios: with 3dfx and sound, crash; with glide passthrough and sound, crash; with sound but without 3dfx, crash; the only scenario that works is when sbtype=none, no sound, then both regular and 3dfx versions play just fine. 

I already discarded a problem with the game because the exact same game/files run just fine on QEMU, both non 3dfx and glide passthrough run just fine there with sound. VMware with a XP VM has the same NFS2SE files and run fine too.

I can reproduce the problem with other games, like FIFA 98 RTWC and FIFA 99. In all tests above, I used the latest available SDL1 version, as well as older builds. I could go as far back as possible to 0.83.7, any older than that my Win98SE wouldn’t complete the boot.

I tried SDL2 yesterday and got the same problem, although when I am on Windows 98SE desktop the screen is yellow/orange tinted.

**Comment by brunocastello** [comment]
Good night, another test report:

DOSBox-X 0.83.18, Apple Silicon M1 MacBook Air 2020, 16GB, 512GB, 8 core CPU, 8 core GPU. Windows 98 SE disk image, with latest SB16 and Voodoo 3dfx drivers. My cfg file isn't VERY different from the recommended one on Win98SE install guide, apart of NE2000 backend selection. No glide, just voodoo emulation (I did not want to mess with my environment this time, trying to build glide libraries, so I decided to try to use a pristine clean environment first).

Trying to run Need For Speed II SE. Game hangs on race loading screen (the one with a track map and loading bar). This was with cputype=pentium. As soon as I moved to cputype=pentium_mmx, I could get to the track for race start (despite of my mouse pointer sometimes freezing for a while intake) but I crash back to the desktop after the green light. Subsequent attempts to run the game did not replicate the same behavior and only stopped at race loading screen.

Same applies to cputype=experimental. No luck. Games are only playable without sound (sbtype=none) :(

**Comment by brunocastello** [comment]
Hi, did anyone investigate the issue? I couldn't find out why these games (and other users reported similar issue with other games) don't crash when sound blaster emulation is not enabled.

**Comment by rderooy** [comment]
Playing around a bit, I found that if I started DOSBox-X with the ``-debug`` switch that I could start NFS2SE multiple times in a row and each time I was able to get to the actual gameplay with sound. No crashes to the desktop.

That to me seems to imply it may be a timing or race issue.

I also tried it with the 3dfx executable, but there the -debug switch causes major slowdown and graphical glitches and it still crashed to the desktop on game start.

**Comment by brunocastello** [comment]
Well, I did list in https://github.com/joncampbell123/dosbox-x/issues/2689#issuecomment-944844853 another three similar issues with sound to help a bit with the problem. One of them were fixed by simply setting cpu to pentium pro, but that did not help in my case.

At first I thought it was a problem with the "emulation speed" because maybe these games were too fast for the "sound emulation" to keep up with it. I was unsure about it. So I tend to accept your theory of a timing or race issue.

In three different platforms (Intel Mac, M1 Mac, Intel Windows) I had the same issue with the same Win98 disk image. I haven't tried under Ubuntu because I failed miserably at trying to install and run DOSBox-X there (not really a skilled linux user).

To sum up all the posts above, be it 3dfx or software mode, be it sb16 or sb16vibra, be it irq 5 or irq 7, whenever I start a race on NFS2SE or a match in FIFA 9x, I have a CTD straight after the sound starts to play. I noticed a bit of a delay when I move the mouse pointer over NFS2SE menu items before starting a race, though. While in FIFA, when the game froze, I can hear the commentary still going on.

QEMU has the same sb16 card and drivers. I used the same image there and had zero problems to play a game there. But there, I can use Pentium II and/or Pentium III. I'm using PIII because Grand Prix 3 needs at least PII.

I understand that the new cpu option pentium_ii in the upcoming next release is not any faster than the current ones, just purely new instructions?

**Comment by rderooy** [comment]
Yes, the Pentium II emulation just adds the additional instructions. If your computer struggles to emulate a Pentium MMX, it will not do better emulating a Pentium 2.

In any case, my understanding is that you cannot play a game of NFS2SE at all without crashing to the desktop. While I can, it only crashes to the desktop if I restart the game. Which leads me to think your system is more susceptible to this issue, which again could be a timing or race issue.

**Comment by brunocastello** [comment]
> Yes, the Pentium II emulation just adds the additional instructions. If your computer struggles to emulate a Pentium MMX, it will not do better emulating a Pentium 2.
> 

Thanks for clarifying that information for me. Luckily, I don't have that problem even with the 2013 rMBP and the 2020 M1 MBA.

> In any case, my understanding is that you cannot play a game of NFS2SE at all without crashing to the desktop. While I can, it only crashes to the desktop if I restart the game. Which leads me to think your system is more susceptible to this issue, which again could be a timing or race issue.

Well, what I can say... all that I know is that I tested it in three different system (an Intel retina MacBook Pro 2013, a brand new M1 MacBook Air 2020, and an Intel Windows 10 x64 virtual machine, on the same Intel MacBook Pro through VMware Fusion). The first thing I did when I got the M1 Air, was to run DOSBox-X to have a test with a completely pristine and brand new environment. No joy. Last night I did another attempt with the last SDL1 release available for M1 (SDL2 also fails the same way) and this time I used the exact same configuration file posted on Win 98 guide, no luck either.

I understand your point about my system being more susceptible, but I tested it with three different environments and all three had the same problem. Which means that it can't be the system environment and has to be something on DOSBox-X, so we are back to square one and still with the theory of a timing or race issue.

I also borrowed another Windows 98 image for DOSBox from a friend, he had a clean install, I installed NFS2SE, FIFA 98 RTWC, FIFA 99 there and had the same problems. He tested himself on his system too, and had the same issue. I also tried to create a new Windows 98 install from scratch, installed all three games and the problem was still there.

I used the same process for QEMU and had zero issues there with the three games. So, these were all the tests (along with the all other tests related on this thread before) I did on this matter.

**Comment by brunocastello** [comment]
Actually, I have another theory but first I need to ask a question: Does the voodoo emulation have any way to set up a Slow or Fast LFB Handler?

**Comment by rderooy** [comment]
If you think the issue is related to the voodoo emulation, just disable it.
```
[voodoo]
voodoo_card=false
```

I tried it here, and was able to restart the game a second time, but on the 3rd restart of the game it crashed back to the desktop (but a bit later than otherwise).

**Comment by brunocastello** [comment]
Even without voodoo it crashed the same way - match kick off scene, and froze.

**Comment by brunocastello** [comment]
@rderooy I see that there is no ARM build for the latest version to test on my M1 Air.

How can I build it myself here to test, but without fluidsynth and without avcodec? 

Intel build is extremely slow on M1 Macs, which is understandable because of Rosetta 2.

**Comment by rderooy** [comment]
I do the Linux x86_64 and arm64 builds (flatpak and rpm). The MacOS builds are done by @joncampbell123

**Comment by joncampbell123** [comment]
@brunocastello I was unable to produce an ARM binary because XCode's tools became unexpectedly neurotic about some __LINKEDIT segment thing and Brew dylib libraries.

If you relied on a command line tool to alter the RPATH entries of a dylib and it suddenly considered it a fatal error for "link edit data" not to fill a __LINKEDIT segment no matter how irrelevant to what you _wanted_ it to do, wouldn't you get pissed off too?

Despite Apple's terrible modern documentation I am gathering file format information on the "MachO" format used in Mac OS X so I may be able to fix that by writing my own command line tool to change the rpath. And unlike Apple XCode, I won't faint and clutch my pearls in horror if some data doesn't fill a segment.

I was able to make an Intel binary because I hadn't yet updated to the latest XCode, so it was able to change rpath without worrying about __LINKEDIT.

Why, yes, I am kind of sour about command line tools that fail me like that.

<img width="1280" alt="Screen Shot 2021-11-30 at 10 45 46 PM" src="https://user-images.githubusercontent.com/6245486/144661736-c69cdb0d-afab-49ad-be3c-ef3a7a4a3897.png">

**Comment by brunocastello** [comment]
Apple being Apple.... *sighs* I completely understand your frustration.

**Comment by joncampbell123** [comment]
@Wengier Is there a way to see the output of the build steps when continuous integration fails?

GitHub is giving me the equivalent of DUH SOMETHING WRONG NO ME NOT TELL YOU WHAT.

![Firefox_Screenshot_2021-12-04T00-24-22 697Z](https://user-images.githubusercontent.com/6245486/144688905-9345e329-b174-4a1e-8efd-689cf782ada3.png)

**Comment by Wengier** [comment]
@joncampbell123 Wow, you are adding ARM64 macOS CI target. There is no workflow posted for this target, but for the standard macOS target you can see the CI output easily, just from the build window:

![image](https://user-images.githubusercontent.com/8216923/144689511-676bda4d-7f23-405e-bb7b-2f7f4650fa1a.png)

Or you can view the raw CI output (in plain text) by clicking "View raw logs" option in the top-right:

![image](https://user-images.githubusercontent.com/8216923/144689602-0dec01e7-7ecf-4fb8-9706-fd07ab9bc04b.png)

Then the complete CI output will be shown there.

**Comment by joncampbell123** [comment]
I didn't add the ARM64 CI target, it was there when I checked it. GitHub emailed me to notify me which CI steps failed.

**Comment by Wengier** [comment]
@joncampbell123 Where did you find the ARM64 macOS CI target? Is there a link to it so I can see as well?

**Comment by joncampbell123** [comment]
This is what I see and---Oh. Never mind, false alarm. :laughing: 

![Firefox_Screenshot_2021-12-04T00-44-09 553Z](https://user-images.githubusercontent.com/6245486/144689843-f0a63a5c-1092-4aea-9a92-cc6009f80e56.png)

GitHub notifies me whenever DOSBox staging does some things.

**Comment by joncampbell123** [comment]
It looks like the guys over at DOSBox staging figured it out. Let's ask them how they did it.

This is what GitHub emailed me:

https://github.com/joncampbell123/dosbox-staging/actions/runs/1533104791

**Comment by Wengier** [comment]
It is from the DOSBox Staging project then. It is probably self-hosted, rather than provided by GitHub.

**Comment by Wengier** [comment]
Yes, it is indeed self-hosted:

![image](https://user-images.githubusercontent.com/8216923/144690134-f96029ed-dded-4bda-b2e7-563ab047b88c.png)

**Comment by brunocastello** [comment]
Ouch! Let me know if you want me to build one here, but without fluidsynth and avcodec. My M1 Air does have a build environment, I had to have it so I could build QEMU-3dfx...

**Comment by joncampbell123** [comment]
@brunocastello You're free to try... though you might get the same error message.

**Comment by brunocastello** [comment]
I actually did try it last week, and it eventually failed. Can't remember why now.

**Comment by grapeli** [comment]
@brunocastello 
I confirm with `sbtype=none` Need For Speed II SE in the demo version (and full) it works fine (as much as).

https://user-images.githubusercontent.com/452325/144729323-e9105e99-6d34-40a3-a9e3-ccdc2496c424.mp4

With `sbtype=sb16` you exit the game after ~ 10 seconds.

https://user-images.githubusercontent.com/452325/144729338-52100860-c700-4f29-9ad1-35e063ed467c.mp4

Of course, this does not apply to the performance of my 2010 laptop with Intel Core Westmere clocked at 2.4 GHz (with turbo 2.6). Please do not suggest yourself.

**Comment by brunocastello** [comment]
Wine sucked for nearly all games I tested on my M1. IMHO, dosbox-x should be the best solution for mac users.

I have been using qemu-3dfx but I am not happy with it either. It’s a half baked solution, I have glide but not mesa passthrough. Needs specially patched wined3d libraries for that.

a cross compiled build of PCem for mac works, but voodoo emulation is broken.

So it leaves us with dosbox-x as the best possible option available at the moment.

**Comment by grapeli** [comment]
@brunocastello 
I am convinced that most of the titles you mentioned in this thread work under wine with gold or platinum status. Of course, you have to know how to handle wine. It also has regressions. It is worth having wine in different versions (from 1.8.x, 3.0.x to the latest 6.23). This NFSIISE demo should be run like this:
`taskset -c 0 wine NFS2SEA.EXE`
For wine, they didn't work well for me:
-Big Red Racing (requires 256 color spaces) - looks awful under wine
-Freddy Pharkas: Frontier Pharmacist (version under Win 3.11) - was launched partially.

Using wine is pure pleasure compared to the torment of installing Windows 98.
Enough of wine.

I am curious how Formula 1 Demo 3DFx (Psygnosis) with `sbtype=sb16` works for you under DOSBox-X.
https://archive.org/download/Formel1/Formel1.zip
It works for me (graphic artifacts are a bit annoying).

https://user-images.githubusercontent.com/452325/144743289-ceb905d8-3695-4194-a7ee-2612df58db46.mp4

**Comment by grapeli** [comment]
After disabling acceleration support for sb16 audio playback in Win98, NFSIISE works fine (except for unwanted graphics artifacts that appear).

https://user-images.githubusercontent.com/452325/144752896-01f8fb1e-d920-4a50-93e9-b347a3f1e373.mp4

**Comment by brunocastello** [comment]
I also tried disabling hw accel support for sb16 in win98. Didn’t help much other than giving me more minutes before another freeze.

As for Wine, I don’t use linux, the options in a M1 MB Air are limited to crossover and wineskin winery on macOS, both don’t work too good. I only managed to run Grand Prix 4 and Counter Strike 1.6. However, I cannot use it; after a long period of play, the mac is overheating more than my previous Intel MB Pro did. That’s because these two games are demanding a lot from emulation, and I am concerned with battery health, so I stopped using it.

I never had problems with dosbox and qemu other than this sound issue on dosbox. M1 temperature remains pretty stable, it does heat a bit but nothing much different to the 2nd gen iPad Pro I previously owned and sold later. 

M1’s thermal temperature works amazingly well, even when I have a 8 hour work on Adobe XD it remains colder than a winter in Spain…

@grapeli try dxdiag and run the sound tests. you’ll see it fail, but the outcome might help us to find more about this problem.

**Comment by grapeli** [comment]
Passes this test.

https://user-images.githubusercontent.com/452325/144758457-97c80371-8e8d-4483-8d2b-01340a509f63.mp4

Defect after pressing "Exit" from dxdiag, an error occurs, dosbox-x exits with the following message:
`E_Exit: RET from illegal descriptor type 0`.

**Comment by grapeli** [comment]
Premature hurrayoptimism - changing disabling hw accel only extends the stability in NFSIISE up to 2 minutes (not enough).

**Comment by brunocastello** [comment]
Mine didn't pass the dxdiag tests. I was gifted with a message about only software modes being supported, no hardware acceleration at all. Even though I have tried with both hw acceleration enabled and disabled.

**Comment by grapeli** [comment]
Passing the test = I can hear the correct playback of individual sound samples.

`You sound card does not support hardware buffering. Sounds will only play back from software buffers.`
This message does nothing to pass the test in my opinion.

**Comment by brunocastello** [comment]
Yup, I know. This was the exact same message I got here.

With QEMU-3dfx I made the same test and it passed the tests, no unexpected messages until the end.

**Comment by joncampbell123** [comment]
Muahaha... good news, I finished an in-tree tool to patch dylib references in Mach-O dylib and executable files.

I am able to make .app bundles for ARM Mac OS X again, by using my own custom tool instead of install_name_tool.

https://github.com/joncampbell123/dosbox-x/blob/master/src/tool/mach-o-matic.cpp

**Comment by brunocastello** [comment]
@joncampbell123 I compiled the latest code and noticed one thing:

The dynamic_x86 option is not available anymore. Only dynamic_rec if I want to use a dynamic core. Is it proposital?

**Comment by grapeli** [comment]
The sound problems are rather related to NFSIISE. The included readme.txt file contains the following information.

>SOUND CARDS WITHOUT DIRECTX SUPPORT
>===================================
>If you are using a sound card, Need For Speed II SE requires that it be supported by DirectX
>drivers. If your sound card drivers do not have DirectX support, you may experience delayed
>sound, stuttering, or sound drop out.  In addition, you may experience some difficulties
>maneuvering in the menus with your joystick, if it is attached to the sound card's game
>port.  If this is the case, you can maneuver in the menus with your mouse or keyboard.

I tested a few other games and there was no similar sound problem with them.

**Comment by joncampbell123** [comment]
@brunocastello dynamic_x86 and dynamic_rec availability is chosen at compile time.

**Comment by brunocastello** [comment]
> The sound problems are rather related to NFSIISE. The included readme.txt file contains the following information.
> 
> > # SOUND CARDS WITHOUT DIRECTX SUPPORT
> > If you are using a sound card, Need For Speed II SE requires that it be supported by DirectX
> > drivers. If your sound card drivers do not have DirectX support, you may experience delayed
> > sound, stuttering, or sound drop out.  In addition, you may experience some difficulties
> > maneuvering in the menus with your joystick, if it is attached to the sound card's game
> > port.  If this is the case, you can maneuver in the menus with your mouse or keyboard.
> 
> I tested a few other games and there was no similar sound problem with them.
> 
> edit: Another such observation. It is stable for max 120 seconds with `sbtype=sb16` and `cputype=pentium_ii` only. With `pentium_mmx`, you exit the game as soon as the track location is loaded.

@grapeli I wouldn't put it all down to NFS2SE when there are other 3 or 4 issues I mentioned in this thread with other games plus more two games I tried (FIFA 98 RTWC and FIFA 99). Let me find them to comment here for reference.

**Comment by brunocastello** [comment]
> @brunocastello dynamic_x86 and dynamic_rec availability is chosen at compile time.

How do I choose it? I tried when I first managed to compile your new code this afternoon, but looks like I did something wrong. I basically edited build-macosx to build an ARM version without avcodec and libfluidsynth (until then I was using the release builds) and tried to enable dynamic_x86. I might have missed something.

**Comment by Wengier** [comment]
@brunocastello The dynamic_x86 core is available on x86 and x64 computers, but not on ARM computers. I think it will not be available if you build the code on an ARM64 computer.

**Comment by brunocastello** [comment]
Hmmm... if I understood it correctly, if I cross compile for M1 on my old Intel MBPro, it won't make any difference at all? Bummer... because the dynamic core on M1 is somewhat slower with the dynamic_rec, but OK.

**Comment by joncampbell123** [comment]
dynamic_x86 is more x86-oriented, by it's name :)

**Comment by brunocastello** [comment]
Well, I plan this weekend to try out a Win 95 install on both DOSBox-X and DOSBox vanilla, to see if the sb16 problem still persists

**Comment by grapeli** [comment]
The NFS2SE version without 3dfx with sb16 is clearly more stable for me. At first I got the impression that it was completely. The first attempt with a race lasting over 5 minutes was completely successful (I did not want to play longer).

During this recording, I quit the game unexpectedly. The audio lag is significant.

https://user-images.githubusercontent.com/452325/145673828-4f765b2d-dd3f-4466-a918-1ffdf110436c.mp4

**Comment by rderooy** [comment]
As I mentioned, after starting Win98SE I can play NFS2SE just fine with sound (I tried 2 races in a row), it is only after I quit the game and re-launch it that it crashes back to the win98 desktop.

But when I messed with the graphics options in NFS2SE and turned everything up to max, after starting the race after a few seconds it dropped me to the desktop.

This is on my Linux system (AMD Ryzen 5600X)

**Comment by DSN9990** [comment]
I recently installed a Win98SE guest just only for testing the new Pentium 3 CPU and some games, and I used software 3Dfx emulation (and sometimes OpenGL 3dfx emulation).

I tested two demanding games, Blood 2 and Unreal (original 1998 version).

For some reason Unreal does not seem to work in normal mode (which will enable 3dfx acceleration if detected) and crashes after initializing the engine's final phase, but works in safe mode (which runs in a window and uses software rendering):


![guest os_006](https://user-images.githubusercontent.com/69558812/152117446-661e4612-739a-4bcd-b895-bf5384597f50.png)

Blood 2 seems to work fine, despite running very slow and with polygon artifacts mentioned in issue #1784 (it occurs when DirectX 6 or higher is installed in the guest, and as such I am using 98 SE since it comes with DX 6.1 out-of-the-box).

As I mentioned Unreal seems to work fine (I guess) in software rendering but not in 3d acceleration mode for some reason. Can @joncampbell123 check this out when possible? I think this issue I am talking about has nothing to do with this thread, but I'd like help with this crash, though. I checked the log file too but it appears to not have an error message related to the crash, however.

**Comment by DSN9990** [comment]
Works fine in safe mode (there is no sound this way):

https://user-images.githubusercontent.com/69558812/152123565-71e91520-67cc-4c3c-8833-b921295591a2.mp4

**Comment by DSN9990** [comment]
Works fine in safe mode (there is no sound this way):

https://user-images.githubusercontent.com/69558812/152124398-f4487cb1-e96b-4adc-a9dd-53faea5de621.mp4

**Comment by brunocastello** [comment]
You can try out Need For Speed II SE and FIFA 99.

If you can race without a crash to desktop several times, and if you can play a match without the game screen freezing on match kick-off while the commentary is still going on, then you've found the gold.

Don't forget to share the configuration in case you manage it. I've been trying that since 2020 without much success.

**Comment by grapeli** [comment]
@brunocastello 
Wine is a first-class platform for running games under linux. I suspect that it also works well under macOS. Only on Windows it does not work well, but when microsoft refines accelerated video output to a linux virtual machine, then it will call it something WSL3, then Windows users will be able to use wine.

https://user-images.githubusercontent.com/452325/152619181-313dea50-ec1c-4b80-b291-a5265d43e295.mp4

NFSIISE works very well under wine. Also in the 3DFx version (maybe even better). In the latter case, a cross-platform wrapper can also be used.
https://github.com/zaps166/NFSIISE

**Comment by brunocastello** [comment]
@grapeli Thank you, I appreciate your suggestion, but my problem here is more than just one game in particular. There are other games like FIFA 9x and NBA Live 98 in the same situation, and I have collected a few reported issues from other users here with a similar problem, and these issues may be pointing the finger to some SB16 emulation issues; when there is no sound device being emulated, these games can run.

I chose to go with a qemu fork with glide passthrough support. Works fine for glide games. Not so well for MESA GL passthrough since I cannot find a proper WineD3D library to do it. And the full screen support is lackluster.

But anyway, it runs nearly all of my games so I am kinda satisfied: https://github.com/kjliew/qemu-3dfx

I know WINE, I am forced to use CrossOver for more modern games like Grand Prix 4, but it's not what I want; I want a full classic windows environment and old games for nostalgia purposes. Running a classic game under a modern environment is not exactly my goal here. DOSBox and QEMU are my only alternatives to do that on my M1 Mac.

**Comment by brunocastello** [comment]
So... has anyone else tried to investigate the problem further?

**Comment by brunocastello** [comment]
As of the latest release .23, the issue still remains persistent. Gaming is only possible with sbtype=none.

**Comment by brunocastello** [comment]
I have just created today a brand new Win 98 install using a different ISO image than I had used previously, and following the same instructions provided by @rderooy and I still cannot game. I installed Need For Speed II SE to test. I can only play the game when sbtype=none, otherwise game crashes back to win 98 desktop when the race starts. I have the latest DX9, latest Voodoo and latest SB drivers.

I am having trouble to diagnose why I have that problem since 2020 on both Intel and M1 Macs, even though I have not done any special stuff on my environment that could cause it. I could only isolate the problem and it is totally related to Sound Blaster emulation within games.

**Comment by Wengier** [comment]
@brunocastello Perhaps you can actually post your Windows 98 image for others to test?

**Comment by brunocastello** [comment]
> @brunocastello Perhaps you can actually post your installed Windows 98 image for others to test?

Its a 16gb disk image. I could upload it to WeTransfer but I’m in Spain with an limited internet. I’ll be back to Rio in May.

Same install files I used for this one works 100% when I use qemu-3dfx, but it’s not practical.

**Comment by grapeli** [comment]
Very small win98se image, 146MB compressed, 400MB unpacked.
https://drive.google.com/file/d/1bu4562Qkj_JKug_0ESrBMSLZOUBCIHPp/view?usp=sharing
Includes three demos - NFS2SE, NBA 98Live, and F1 Formula (Psygnosis). All with 3DFx support.

I run like this.
`dosbox-x -fastlaunch -set "log logfile=dosbox-x.log" -set "log vga=debug" -set "log vgagfx=debug" -set "log vgamisc=debug" -set "log int10=debug" -set "log dma_control=debug" -set "log fpu=debug" -set "log cpu=debug" -set "log paging=debug" -set "log fcb=debug" -set "log files=debug" -set "log ioctl=debug" -set "log exec=debug" -set "log dosmisc=debug" -set "log pit=debug" -set "log keyboard=debug" -set "log pic=debug" -set "log bios=debug" -set "log misc=debug" -set "log io=debug" -set "log pci=debug" -set "log sst=debug" -set "log int21=true" -set "log fileio=true" -set aspect=true -set "dos lfn=true" -set "dos ver=7.10" -set "dosbox quit warning=false" -set "dos hard drive data rate limit=0" -set memsize=128 -set showmenu=false -set char9=false -set scaler=none -set cputype=pentium -set core=dynamic_x86 -set output=opengl -set voodoo_card=opengl -set cycles="max 105%" -set "ide, primary int13fakeio=true" -set "ide, primary int13fakev86io=true" -c "@imgmount c win98se.img >NUL" -c "boot c:"`

NFS2SE is very unstable (`sbtype=sb16` sound enabled). Works from 1s to 4 minutes correctly.
Stability definitely decreases with `cputype=pentium_ii`. Run with `cputype` 486 or pentium.

Unfortunately, there are some unwanted graphic artifacts in each of the demos. They are the least important in the NBA, because only on the menu. In the F11, they are visible in front of the car's hood practically all the time.

**Comment by brunocastello** [comment]
I just compiled the latest source code (0.83.25) with the PR 3428 to test the SDL2 yellow tinted windows fix.

The yellow tinted windows was fixed. However, when I try to run a 3dfx game (like NFS II SE) DOSBox-X crashes and closes.

I tried the non-3dfx version of NFS II SE and it does run.

**Comment by brunocastello** [comment]
@joncampbell123 as we spoke on Discord yesterday, I ran a few tests. I haven't tested a build compiled without FFMPEG yet.
But last build from 2 days ago had --disable-libfluidsynth and --disable-avcodec. So I think this test has been done before, right?

Anyway, the latest tests. I installed Win 98 SE fresh, updated drivers (both sb16 and voodoo, and installed VBEMP 9x drivers for higher resolutions*). And installed Need For Speed II SE as usual. Then I proceed to test the 3Dfx version of the game, the reason why I want to run it on dosbox-x.

0.83.25 and 0.83.24 = NFS II SE crashes dosbox-x and closes it when I attempt to run the game
0.83.23 and 0.83.22 = I can start the game, go through RACE menu, and as soon as I get to the track, the game crashes back to Win 98 desktop. Subsequent attempts to run the game again have the same fate.

I am yet to try out without sound and to try out without FFMPEG. I have a barbecue to go now, in a few hours I'll be back to run the remaining tests.

One thing to note, is that when I get to the main menu, there are a few graphics glitches up and down. When I attempted to use a ready-made Win 98 install from QEMU, this did not happen, however I couldn't race too. That install had nothing different, same drivers, same files.

* = I had even tested without that driver before, and the problem persisted.

**Comment by brunocastello** [comment]
I had a few minutes to test without sound (sbtype=none) before going out.

**0.83.21:** game runs just fine without sound.
**0.83.23:** game runs just fine without sound.
**0.83.24:** (current release, April 1) game crashes when attempt to run it. This release is also missing the SDL2 yellow screen fix that I tested a few days ago.

Next when I'm back is to test compiling a SDL2 build without FFMPEG (--disable-avcodec, right?) However, I don't expect it to be the problem, because I have compiled with that flag in the past year and the problem persisted.

**Comment by maron2000** [comment]
NFS2SE runs fine but without sound & Voodoo on 0.83.25 Windows x64 SDL1/SDL2 Visual studio build.
I finished a two laps race.
CPU is Dynamic core / auto.

**Comment by brunocastello** [comment]
Yeah, macOS version of Dosbox-X has a problem on 0.83.25. The emulator just crashes whenever I run the game.

Attached the dosbox-x.conf I am using (had to save as txt to be able to upload):
[dosbox-x.txt](https://github.com/joncampbell123/dosbox-x/files/8602318/dosbox-x.txt)

I want to run certain settings to be able to run things like Netscape 9 (heaps of memory) and some certain games, not just NFSIISE.

**Comment by grapeli** [comment]
Linux DOSBox-X 0.83.25 SDL2.
Works fine with 3dfx without sound. There are no unwanted graphic artifacts. It is fully stable.

double speed of the video

https://user-images.githubusercontent.com/452325/166226760-434cc34b-5f41-4e72-b519-f869539cdde3.mp4

With `sbtype=sb16` it is unstable. Unwanted graphic artifacts appear. Audio delay is up to 20 seconds (after leaving the tunnel).

https://user-images.githubusercontent.com/452325/166226869-17466df9-feae-4d89-81d7-9c8036274dd7.mp4

There is no difference in operation between the demo and the full game. Both versions behave identically.

**Comment by brunocastello** [comment]
@joncampbell123 this is the error report generated when I try to run NFS2SE with a 0.83.25 build and it crashes.

[report.txt](https://github.com/joncampbell123/dosbox-x/files/8603331/report.txt)

I've just compiled now this build with --disable-libfluidsynth (because the build script installs fluid-synth from brew with some other dependencies) and --disable-avcodec (because you wanted to see something about FFMPEG).

**Comment by joncampbell123** [comment]
> @joncampbell123 this is the macOS error report generated when I try to run NFS2SE with a 0.83.25 build and it crashes.
> 
> [report.txt](https://github.com/joncampbell123/dosbox-x/files/8603331/report.txt)
> 
> I've just compiled now this build with --disable-libfluidsynth (because the build script installs fluid-synth from brew with some other dependencies) and --disable-avcodec (because you wanted to see something about FFMPEG).

It crashed in glViewport? Figures. I had to make other adjustments some time ago when calling glFinish before starting a frame would crash the Mac OS X OpenGL library if no prior frame had been started.

There is already a flag to track that for that reason, perhaps that flag needs to be used in the Voodoo emulation as well, not to call certain functions until a proper frame has been started.

**Comment by joncampbell123** [comment]
Here on Linux I can confirm non-accelerated NFS2SE will eventually just dump back to the desktop without any error message.

I've thought about one possible cause, and it has to do with the audible artifacts when Windows 95 starts up. DOSBox-X inherits DOSBox SVN's millisecond-by-millisecond Sound Blaster Processing and the "minimum DMA" processing that I can only assume was meant to keep games from stuttering.

If NFS2SE is using DirectX and DirectSound to render audio, perhaps it's sensitive to any "jitter" that might happen from the 1ms-based processing. Having written code to deal with DirectX, DirectSound primary buffers are either emulated as or actually handled as a circular DMA buffer, just like you would do in MS-DOS with Sound Blaster. The theory is that maybe NFS2SE has naive code that assumes the DMA will always advance monotonically, and if it judders back a bit because of some naive handling by the sound driver, the game computes some extra large integer and throws an assertion check or error function, or maybe tries to render that many samples and throws an exception.

I suppose one way to verify that would be to whip up a Win32 program that starts a DirectSound primary buffer and then continually watches and reports it's movement to a file.

**Comment by joncampbell123** [comment]
To explain the programming mistake, my code back then used DirectSound to record from the microphone at 44.1KHz 16-bit stereo. I was writing something to record radio shows and I had arranged a sound board to use two microphones side by side to pick up the voice in stereo.

It would have been a good test recording except that I found out too late the L and R channels constantly switched back and forth.

My code assumed the playback byte position reported would be aligned to a complete stereo sample in the buffer, but because the Windows 98 drivers reported directly from the DMA controller, the reported byte position was sometimes in the middle of the complete sample, meaning, since it was obviously using 16-bit DMA, sometimes the reported position would be right between the 16-bit L and R samples aka the DMA had written L, but not yet R. I've been writing my code to round down the reported position by nBlockAlign since.

**Comment by joncampbell123** [comment]
A quick check: Whatever it doesn't like about SB16 emulation it dislikes the ESS 688 AudioDrive emulation even more.

**Comment by casasfernando** [comment]
Just found this github issue after troubleshooting the problem for a few hours.
I’m using Dosbox-X flatpak latest version (2022.09.0 64bit sdl2) and indeed changing sbtype to none stops NFS2SE from crashing anymore (3dfx enabled).
I can also confirm that the same issue affect other NFS games like High Stakes.

**Comment by grapeli** [comment]
Under dosbox-pure it is [very similar](https://github.com/schellingb/dosbox-pure/issues/279). I suspect that if someone would test NFS2SE under dosbox-svn or dosbox-staging, of course, with the necessary patches extending the capabilities of these programs, the effect would be the same. 

The error is hidden deep in the dosbox code.

I don't think it has anything to do with soundblaster. I tested NFS2SE under dosbox-x with GUS and it behaves the same as with soundblaster. There is a quick exit when the game starts.

With sbtype=sb16 and 3dfx I sometimes managed to complete the whole race (over 4 minutes), but most of the time the error occurred 15-30 seconds after the start of the race.

**Comment by casasfernando** [comment]
@grapeli Thanks for sharing this as testing GUS was my next step so I can skip it now.
I’m wondering how can this be debugged because certainly not from the game itself.
Maybe from DirectX if this is what for instance NFS2SE uses for sound?

**Comment by grapeli** [comment]
@casasfernando 
As I thought this bug is not exclusive to DOSBox-X. It's the same with dosbox-staging. I tested on the latest version git with five patches from the [ww/fat32-test-1](https://github.com/dosbox-staging/dosbox-staging/tree/ww/fat32-test-1) and [ww/voodoo-3dfx-2](https://github.com/dosbox-staging/dosbox-staging/tree/ww/voodoo-3dfx-2) branches included.

In the recording, the exit to the desktop occurs very quickly. Other times it took about a minute and a half.

[NFS2SE.dosbox-staging.webm](https://user-images.githubusercontent.com/452325/202015068-2852ae2f-483a-481a-8670-c0114962635a.webm)

**Comment by casasfernando** [comment]
@grapeli in my case with none of the games I get that far, meaning race shows for a few seconds. For me, both NFS2SE and NFSHS exit one second or less after the race loading finishes and the "race environment" is displayed. Sometimes I get to see a picture of the care or the track, but it crashes almost after a seconds.

Anyways, the problem is how to debug this issue. I'm willing to help testing, collecting logs, etc. if guidance is provided on what information would be needed.

Thanks

**Comment by grapeli** [comment]
If this bug is so hard to diagnose with NFS2SE it might be worth a try with the [Need for Speed High Stakes Demo](https://archive.org/details/NeedForSpeedHighStakesDemo), which behaves very similarly, except that exiting the game to the desktop ends with a detailed error message.

![guest os_000](https://user-images.githubusercontent.com/452325/202292226-b0ef9c4f-879c-4d7c-8dd0-fde7b2451b4a.png)

**Comment by grapeli** [comment]
The error `EXCEPTION_ACCESS_VIOLATION: read attempted` in High Stakes appeared on real hardware configurations.
The most common cause suggested is a problem with the graphics driver or the directx version.
You can find this information in the `readme.txt` file for this demo.

>AUDIO PROBLEMS
>In some cases, sound problems are caused by video drivers. Make sure you have the latest video drivers for your video card.

Of course, this demo with `sbtype=none` works perfectly fine. Except for one drawback - there is no sound.

**Comment by casasfernando** [comment]
@grapeli which DirectX version and graphics drivers are you using?
In my case I installed (maybe I shouldn’t) DirectX 9.0c and the Voodoo drivers are the one mentiones in the wiki for Win98. I can check the exact version later on and report back.

**Comment by grapeli** [comment]
@casasfernando 
I posted a link to the win98se disk image a few posts above. I'm testing it on exactly the same.

![guest os_000](https://user-images.githubusercontent.com/452325/202445203-59c13e5a-07d8-43a9-8d49-4ba63e3f4407.png)
![guest os_001](https://user-images.githubusercontent.com/452325/202445206-37a67c45-3f2c-4214-9250-53a84b0bc29f.png)
![guest os_002](https://user-images.githubusercontent.com/452325/202445208-0412a722-49d8-499b-b9d0-6e01486322b9.png)

**Comment by casasfernando** [comment]
@grapeli my setup is mostly the same except that I'm using the S3 emulated card with the OS driver. I'm wondering if it makes sense trying with an older version of DirectX. I was planning to create a new image anyway because wanted to switch Win98 to english (currently I'm using the spanish version) so when I do that I will try to test NFS2SE and NFSHS with whatever version comes with the game and check if it makes any difference but I'm not holding my breath.

**Comment by grapeli** [comment]
@casasfernando 
I haven't noticed a difference in stability between the two drivers. Also in those games listed here.

The main and essential reason why I use the **VBE driver is its much better performance**.

Run [Crystal Mark](https://crystalmark.info/download/archive/CrystalMark/CrystalMark2004R7.zip) on both - CM09GDI and CM09D2D. You can read the results in ini files. While in GDI the difference is tiny, in 2D it is 50% in favor of VBE. Huge. This difference is also visible in other cases I tested.

**Comment by kcgen** [comment]
@grapeli , 

I understand Staging's Voodoo and FAT32 branches are ports of the DOSBox-X sources. Maybe Meson's Address and Undefined Behavior Sanitizer (UB+ASAN) build will reveal issues in the Voodoo or FAT32 code?

Setup:
```
meson setup -Dbuildtype=debug -Db_sanitize=address,undefined -Doptimization=0 -Db_lundef=false -Dc_args=-fsanitize-recover=all -Dcpp_args=-fsanitize-recover=all -Dunit_tests=disabled build/clang-ubasan
```

Build:
```
meson compile -C build/clang-ubasan
```

Executable: `build/clang-ubasan/dosbox`

It will run slow, but it might help reveal issues at the moment the game cuts out back to Windows, which would be common to both projects.  Atleast, if the code has problems at the host level (as opposed to emulation level).

**Comment by grapeli** [comment]
No chance to run Win98SE in dosbox-staging built with clang sanitizer (llvmorg-16-init-10467-g1239d37b-1).
Unless I'm doing something wrong.

Dosbox-staging terminates shortly after the win 98 desktop appears. There is still a long way to go before the system is fully loaded.
Log with `core=dynamic`. 
[dosbox-staging.clang-ubsan.core-dynamic.log](https://github.com/joncampbell123/dosbox-x/files/10040898/dosbox-staging.clang-ubsan.core-dynamic.log)
Log with `core=normal`.
[dosbox-staging.clang-ubsan.core-normal.log](https://github.com/joncampbell123/dosbox-x/files/10040899/dosbox-staging.clang-ubsan.core-normal.log)

In the latter case, once the loading of the system stopped at this point.
![boot_000](https://user-images.githubusercontent.com/452325/202695833-0d656621-c035-466f-80ea-b231080d530f.png)

**Comment by grapeli** [comment]
In dosbox-x compiled with clang (`-fsanitize=undefined`) NFS2SE it starts.
At the time of the error and when exiting the game in the logs I received this information.

```
voodoo_emu.cpp:2635:29: runtime error: index 768 out of bounds for type 'UINT32[9]' (aka 'unsigned int[9]')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior voodoo_emu.cpp:2635:29 in
mixer.cpp:359:62: runtime error: left shift of negative value -20066
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior mixer.cpp:359:62 in
voodoo_data.h:918:46: runtime error: left shift of negative value -3
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior voodoo_data.h:918:46 in
voodoo_opengl.cpp:226:31: runtime error: signed integer overflow: 163 * -9223372036854775807 cannot be represented in type 'INT64' (aka 'long')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior voodoo_opengl.cpp:226:31 in
voodoo_opengl.cpp:226:57: runtime error: signed integer overflow: -931 * -9223372036854775807 cannot be represented in type 'INT64' (aka 'long')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior voodoo_opengl.cpp:226:57 in
voodoo_opengl.cpp:245:35: runtime error: signed integer overflow: 163 * -9223372036854775807 cannot be represented in type 'INT64' (aka 'long')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior voodoo_opengl.cpp:245:35 in
voodoo_opengl.cpp:245:64: runtime error: signed integer overflow: -931 * -9223372036854775807 cannot be represented in type 'INT64' (aka 'long')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior voodoo_opengl.cpp:245:64 in
voodoo_data.h:918:46: runtime error: left shift of negative value -15
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior voodoo_data.h:918:46 in
voodoo_opengl.cpp:226:50: runtime error: signed integer overflow: -9223372036854775547 + -576460752303423280 cannot be represented in type 'INT64' (aka 'long')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior voodoo_opengl.cpp:226:50 in
voodoo_opengl.cpp:245:57: runtime error: signed integer overflow: -9223372036854775547 + -576460752303423280 cannot be represented in type 'INT64' (aka 'long')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior voodoo_opengl.cpp:245:57 in
voodoo_opengl.cpp:226:24: runtime error: signed integer overflow: -9223372036854775807 + -576460752303423234 cannot be represented in type 'INT64' (aka 'long')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior voodoo_opengl.cpp:226:24 in
voodoo_opengl.cpp:245:28: runtime error: signed integer overflow: -9223372036854775807 + -576460752303423234 cannot be represented in type 'INT64' (aka 'long')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior voodoo_opengl.cpp:245:28 in
LOG: VOODOO: OpenGL: quit
LOG: 2295202800 DEBUG MISC:GFX_SetSize 640x480 flags=0x14e scale=1.000x1.000
LOG: menuScale=1
```

I will mention that I previously tested NFS2SE in dosbox-x with both openglide and `voodoo_card=opengl` and `voodoo_card=software`. Same error everywhere. The worst is with openglide, because it freezes dosbox-x.

**Comment by casasfernando** [comment]
@grapeli the above looks like a very good lead on where in the code the problem may be.
What is a bit misleading is that only Voodoo emulation related code is mentioned while the issue seems to be triggered by sound emulation. I’m not familiar with the code so looking forward @joncampbell123 or other devs thoughts on this.

**Comment by grapeli** [comment]
@casasfernando 
It's not so obvious. Please note that this instability also occurs in the non-3Dfx version (`nfs2sen`). This is not exclusive to the 3Dfx version.
I'm not a developer, I don't know what this clang tinkers with when running the code and whether these reports are false.

**Comment by grapeli** [comment]
@casasfernando 
Regarding the non-accelerated version of NFS2SE, it is generally much more stable. Many times I finished the whole race lasting about 6-7 minutes.
I also tested its behavior in dosbox-staging now. I completed two full races totaling nearly 13 minutes. During the third, after about a minute, there was an unexpected end to the game.
Compared to the 3Dfx version, it is much more stable.

**Comment by kcgen** [comment]
Thanks for those build and test results, @grapeli.
This will be useful when we lift up @Wengier's nice voodoo branch to PR - and if we can make any improvements, no doubt we can try them in X and see if helps.

**Comment by maron2000** [comment]
According to @grapeli 's log, there is a left bit shift of a negative value in `mixer.cpp`, which its behavior is undefined if I understand correctly.
Maybe we should change the code not to use bit shift.

**Comment by grapeli** [comment]
@maron2000 
With this setting (`disable filtering=true`), you don't get this mixer warning, and the game still doesn't work properly.

gcc 12.2.0 `-O0 -fsanitize=address -fsanitize=undefined...`

```
LOG: opengl: I am able to use OpenGL to emulate Voodoo graphics
voodoo_emu.cpp:1106:33: runtime error: signed integer overflow: -562949953421312 * -562949953421312 cannot be represented in type 'long int'
voodoo_emu.cpp:1106:83: runtime error: signed integer overflow: -562949953421312 * -562949953421312 cannot be represented in type 'long int'
voodoo_emu.cpp:1107:33: runtime error: signed integer overflow: -562949953421312 * -562949953421312 cannot be represented in type 'long int'
voodoo_emu.cpp:1107:83: runtime error: signed integer overflow: -562949953421312 * -562949953421312 cannot be represented in type 'long int'
voodoo_emu.cpp:1106:33: runtime error: signed integer overflow: -562949953421312 * -562949953421312 cannot be represented in type 'long int'
voodoo_emu.cpp:1106:83: runtime error: signed integer overflow: -562949953421312 * -562949953421312 cannot be represented in type 'long int'
voodoo_emu.cpp:1107:33: runtime error: signed integer overflow: -562949953421312 * -562949953421312 cannot be represented in type 'long int'
voodoo_emu.cpp:1107:83: runtime error: signed integer overflow: -562949953421312 * -562949953421312 cannot be represented in type 'long int'
voodoo_opengl.cpp:253:13: runtime error: signed integer overflow: 13129440 * -9223372036854775807 cannot be represented in type 'long int'
voodoo_opengl.cpp:254:13: runtime error: signed integer overflow: 13129440 * -9223372036854775807 cannot be represented in type 'long int'
voodoo_opengl.cpp:246:35: runtime error: signed integer overflow: 49 * -9223372036854775807 cannot be represented in type 'long int'
voodoo_opengl.cpp:246:28: runtime error: signed integer overflow: -9223372036854775807 + -576460752303423485 cannot be represented in type 'long int'
voodoo_opengl.cpp:246:64: runtime error: signed integer overflow: 451 * -9223372036854775807 cannot be represented in type 'long int'
voodoo_opengl.cpp:247:35: runtime error: signed integer overflow: 49 * -9223372036854775807 cannot be represented in type 'long int'
voodoo_opengl.cpp:247:28: runtime error: signed integer overflow: -9223372036854775807 + -576460752303423485 cannot be represented in type 'long int'
voodoo_opengl.cpp:247:64: runtime error: signed integer overflow: 451 * -9223372036854775807 cannot be represented in type 'long int'
voodoo_opengl.cpp:246:9: runtime error: signed integer overflow: -9223372036854775803 + -576460752303423462 cannot be represented in type 'long int'
voodoo_opengl.cpp:247:9: runtime error: signed integer overflow: -9223372036854775803 + -576460752303423462 cannot be represented in type 'long int'
voodoo_emu.cpp:2679:51: runtime error: index 256 out of bounds for type 'unsigned int [9]'
voodoo_opengl.cpp:221:57: runtime error: signed integer overflow: 372 * -5890867 cannot be represented in type 'int'
voodoo_opengl.cpp:222:57: runtime error: signed integer overflow: 372 * -5890867 cannot be represented in type 'int'
voodoo_opengl.cpp:223:57: runtime error: signed integer overflow: 372 * -5890867 cannot be represented in type 'int'
voodoo_emu.cpp:2635:51: runtime error: index 768 out of bounds for type 'unsigned int [9]'
voodoo_opengl.cpp:226:31: runtime error: signed integer overflow: 4022 * -9223372036854775807 cannot be represented in type 'long int'
voodoo_opengl.cpp:226:57: runtime error: signed integer overflow: 4130 * -9223372036854775807 cannot be represented in type 'long int'
voodoo_opengl.cpp:245:35: runtime error: signed integer overflow: 4022 * -9223372036854775807 cannot be represented in type 'long int'
voodoo_opengl.cpp:245:64: runtime error: signed integer overflow: 4130 * -9223372036854775807 cannot be represented in type 'long int'
voodoo_data.h:918:46: runtime error: left shift of negative value -15
voodoo_data.h:918:46: runtime error: left shift of negative value -3
LOG: VOODOO: OpenGL: quit
```

**Comment by maron2000** [comment]
@grapeli Thanks for testing. There are still bit shifting of negative values and out of bound indexs in the voodoo code, so they have to be fixed.
How about trying the build with `fsanitize=...` without voodoo since there has to be flaws without it as well?

**Comment by grapeli** [comment]
@maron2000 
The Nfs2se demo is only in the 3dfx version.

I tried several times to run the non-accelerated full version of the game under dosbox-x compiled with clang and gcc (-fsanitize=....). Without success. The game starts, after loading the race route, it stops at some point. Most often there is an error in the dosbox-x logs - triple fault.

**Comment by grapeli** [comment]
With `sbtype=none` the demo runs under dosbox-x compiled with gcc (`-fsanitize=...`). Some of these runtime errors related to voodoo appear (mostly when the 3dfx logo shows up).
[nfs2se-demo-gcc-fsanitize.log](https://github.com/joncampbell123/dosbox-x/files/10049257/nfs2se-demo-gcc-fsanitize.log)
My conclusion, this is not the essence of the problem.

**Comment by maron2000** [comment]
@grapeli 
Thank you for sharg iyour research. Although it may not be the cause of this issue, it is a big implication to fix voodoo code.

**Comment by skipster1337** [comment]
On the issue above it's mentioned that disabling sound hardware acceleration supposedly fixes it, but I tried it in DOSBox-X and the game crashed in the middle of the race instead of in the beginning, and also the sound disappears when approaching tunnels. I hope it's possible to fix this issue since DOSBox-X runs games much faster than PCEm and 86Box.

**Comment by DSN9990** [comment]
@joncampbell123 When will this issue be finally fixed? I can't wait for it to be fixed anyway

**Comment by DmytroM1998** [comment]
I'm having the same problem like brunocastello described in 2021. Not just with NFS 2 (regular or special) and Outcast, but with games in general.

I noticed that some games start working more stably if I use `cputype=auto`, but this doesn't help in case of NFS 2. The sound disappears as soon as I start driving. With regular cputype, like `pentium`, the sound is present, but the game freezes in 1-2 minutes of racing. 
In both cases, audio settings don't matter. I read joncampbell's comments about sound. Well, even Mono sound doesn't help.
More cycles doesn't help either.

Windows 10, Ryzen 7 3800X
